osf/models/preprint.py

```diff
@@ -0,0 +1,920 @@
+            if not (self.primary_file and self.primary_file.target == self):
+                raise ValueError('Preprint is not a valid preprint; cannot publish.')
+            if not self.provider:
+                raise ValueError('Preprint provider not specified; cannot publish.')
+            if not self.subjects.exists():
+                raise ValueError('Preprint must have at least one subject to be published.')
+            self.date_published = timezone.now()
+            # For legacy preprints, not logging
+            self.set_privacy('public', log=False, save=False)
+
+            # In case this provider is ever set up to use a reviews workflow, put this preprint in a sensible state
+            self.machine_state = DefaultStates.ACCEPTED.value
+            self.date_last_transitioned = self.date_published
+
+            self.add_log(
+                action=PreprintLog.PUBLISHED,
+                params={
+                    'preprint': self._id
+                },
+                auth=auth,
+                save=False,
+            )
+            self._send_preprint_confirmation(auth)
+
+        if save:
+            self.save()
+        self.update_search()
+
+    def set_preprint_license(self, license_detail, auth, save=False):
+        license_record, license_changed = set_license(self, license_detail, auth, node_type='preprint')
+
+        if license_changed:
+            self.add_log(
+                action=PreprintLog.CHANGED_LICENSE,
+                params={
+                    'preprint': self._id,
+                    'new_license': license_record.node_license.name
+                },
+                auth=auth,
+                save=False
+            )
+
+        if save:
+            self.save()
+        self.update_search()
+
+    def set_identifier_values(self, doi, save=False):
+        self.set_identifier_value('doi', doi)
+        self.preprint_doi_created = timezone.now()
+
+        if save:
+            self.save()
+
+    def save(self, *args, **kwargs):
+        first_save = not bool(self.pk)
+        saved_fields = self.get_dirty_fields() or []
+        old_subjects = kwargs.pop('old_subjects', [])
+        ret = super(Preprint, self).save(*args, **kwargs)
+        if (not first_save and 'is_published' in saved_fields) or self.is_published:
+            enqueue_postcommit_task(on_preprint_updated, (self._id,), {'old_subjects': old_subjects}, celery=True)
+
+        if saved_fields:
+            self.on_update(first_save, saved_fields)
+
+        if first_save:
+            self._set_default_region()
+            self.update_group_permissions()
+
+            self.add_log(
+                action=PreprintLog.CREATED,
+                params={
+                    'preprint': self._id
+                },
+                auth=Auth(user=self.creator),
+                save=False,
+            )
+            self._add_creator_as_contributor()
+        return ret
+
+    def _set_default_region(self):
```

Comment:
Preprints are given the same region as the creator.  Legacy preprints are given the region on the node.  Since preprints don't have addons the region is stored directly on the preprint model.